[quantization] Dont use SubgraphRewriter in FoldQuantizeCallIntoBuffer

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30199 [quantization] Use ConvPackedParams everywhere
* #30198 [quantization] Use LinearPackedParams everywhere
* #30188 [quantization] Memoize parseIR calls in graph mode quantization
* **#30264 [quantization] Dont use SubgraphRewriter in FoldQuantizeCallIntoBuffer**

Differential Revision: [D18645531](https://our.internmc.facebook.com/intern/diff/D18645531)

Commit:
[quantization] Dont use SubgraphRewriter in FoldQuantizeCallIntoBuffer

[ghstack-poisoned]

jerryzh168 commented:
why don't we want to use subgraph rewriter?

jerryzh168 commented:
sorry, misclick

jamesr66a commented:
@jerryzh168 a few reasons:

1. It does redundant work in re-parsing the pattern IR, see https://github.com/pytorch/pytorch/pull/30188#discussion_r349318025
2. It doesn't preserve source ranges, e.g. https://github.com/pytorch/pytorch/issues/30182

jerryzh168 commented:
This complicates the code a lot, is this performance critical? I feel it's better to keep the code simple than to make it more complicated if the perf gain is marginal.
For source range info, can we change the subgraph rewriter to preserve it?

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -948,6 +948,8 @@ graph(%self, %scale, %zero_point, %dtype):
         zero_point_node->kind() == prim::Constant &&
         dtype_node->kind() == prim::Constant;
   };
+  std::unordered_set<Node*> quantize_nodes_to_delete;
```
Can you combine these two sets?

jamesr66a replied:
I can just use a vector to preserve the topological ordering

jamesr66a replied:
Actually if I just reuse the `GetAttr` node we can just avoid deleting those altogether

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -962,15 +964,34 @@ graph(%self, %scale, %zero_point, %dtype):
+  // value. Once we've deleted the quantize nodes, we can delete
+  // the original `GetAttr` nodes.
+  for (Node* n : quantize_nodes_to_delete) {
+    n->destroy();
```
If the reason to split into two sets was to avoid removing a node before all its uses are removed, then you can do it like this:
```
  for (auto n : nodes_to_delete_) {
    n->removeAllInputs();
  }
  for (auto n : nodes_to_delete_) {
    n->destroy();
  }
```

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -962,15 +964,34 @@ graph(%self, %scale, %zero_point, %dtype):
         "_quantized_weight",
         at::quantize_per_tensor(float_weight, scale, zero_point, dtype));
+
+    // Replace the GetAttr[weight]->quantize_per_tensor sequence
```
You can reuse the getattr node and just change the 'name' in it to point to "quantized_weight".

jamesr66a commented:
@jerryzh168 if we want to bias for simplicity, I think the correct direction to move is to enable the SubgraphRewriter to take a lambda for installing custom logic on replacement. That way we can pass in the logic to add the submodule in this use-case (and the others where we peel out the matching functionality)

jamesr66a commented:
> @jerryzh168 if we want to bias for simplicity, I think the correct direction to move is to enable the SubgraphRewriter to take a lambda for installing custom logic on replacement. That way we can pass in the logic to add the submodule in this use-case (and the others where we peel out the matching functionality)

This solution still wouldn't solve the SourceRange issue, though

Commit:
Update on "[quantization] Dont use SubgraphRewriter in FoldQuantizeCallIntoBuffer"


Differential Revision: [D18645531](https://our.internmc.facebook.com/intern/diff/D18645531)

[ghstack-poisoned]

jerryzh168 commented:
@jamesr66a that sounds good, and should be easy to do as well I think?
Why can't we have source range in subgraph rewriter?

jerryzh168 commented:
I know right now we don't support that, but can we improve subgraph rewriter to include source range?

jamesr66a commented:
@jerryzh168 we would just need to figure out some way to correspond nodes between the pattern graph and the replacement graph to do the proper transfer of source ranges. Or some more lossy transfer

ZolotukhinM commented:
> > @jerryzh168 if we want to bias for simplicity, I think the correct direction to move is to enable the SubgraphRewriter to take a lambda for installing custom logic on replacement. That way we can pass in the logic to add the submodule in this use-case (and the others where we peel out the matching functionality)
> 
> This solution still wouldn't solve the SourceRange issue, though

That would perfectly solve the SourceRange issue because in that lambda you can properly set the source ranges for the new nodes. That was a solution for the source ranges I've been thinking about actually, but I wasn't sure it was worth the effort.

ZolotukhinM commented:
> @jerryzh168 we would just need to figure out some way to correspond nodes between the pattern graph and the replacement graph to do the proper transfer of source ranges. Or some more lossy transfer

It is not possible in a general case. The new graph can have nodes without counterparts in the original graph, and there is no way we can automatically deduce proper source ranges for it. We can set the proper source ranges only for inputs and outputs of the subgraph, which remains the same before and after the replacement. That was a half-solution I've been thinking about too, but it will not help for the intermediate values.
Another alternative I considered was to "bind" values by their name in the pattern and replacement subgraphs, but I don't like that because it adds some implicit semantics to otherwise very clearly defined API.

ZolotukhinM reviewed (approved):
LGTM!
